solarforecastarbiter/metrics/probabilistic.py

```diff
@@ -0,0 +1,210 @@
+
+    Returns
+    -------
+    score : float
+        The Brier Score.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    BS = np.mean((f - o) ** 2)
+    return BS
+
+
+def brier_skill_score(fx, fx_prob, ref, ref_prob, obs):
+    """Brier Skill Score (BSS).
+
+        BSS = 1 - BS_f / BS_ref
+
+    where BS_fx is the evaluated forecast and BS_ref is a reference forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    ref : (n,) array_like
+        Reference forecast (physical units).
+    ref_prob : (n,) array_like
+        Probability [%] associated with the reference forecast.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    skill : float
+        The Brier Skill Score [-].
+
+    """
+    bs_fx = brier_score(fx, fx_prob, obs)
+    bs_ref = brier_score(ref, ref_prob, obs)
+    return 1.0 - bs_fx / bs_ref
+
+
+def reliability(fx, fx_prob, obs):
+    """Reliability (REL) of the forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    REL : float
+        The reliability of the forecast, where a perfectly reliable forecast
+        has REL = 0.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities
+    if len(f) < 1000:
+        n_decimals = 1
+    else:
+        n_decimals = 2
+
+    f = np.around(f, decimals=n_decimals)
+    REL = 0.0
+    for f_i in np.unique(f):
+        N_i = len(f[f == f_i])  # no. forecasts per set
+        o_i = np.mean(o[f == f_i])    # mean event value per set
+        REL += N_i * (f_i - o_i) ** 2
+
+    REL = REL / len(f)
+    return REL
+
+
+def resolution(fx, fx_prob, obs):
+    """Resolution (RES) of the forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    RES : float
+        The resolution of the forecast, where higher values are better.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities
+    if len(f) < 1000:
+        n_decimals = 1
+    else:
+        n_decimals = 2
+
+    f = np.around(f, decimals=n_decimals)
```

Comment:
refactor into function, also replace in `reliability`.

Reply:
I'll refactor this into a separate function and add tests (which should also fix the code coverage issue above).

Reply:
might not be needed if we combine the implementations for `reliability` and `resolution`

Reply:
From a testing standpoint, it's easier/clearer to test a separate function (see `_unique_forecasts()` in the current commit). But there should be a way to test the `n_decimals` switch without a separate function (I just need to think about it a bit to figure out a clear way to test).

Reply:
you're right about testing. I have no objection to a separate function, I was just saying it was less important if the code wasn't going to be duplicated.

Reply:
I've started figuring out a way to test the rounding effects without a separate function (i.e. I've confirmed it's do-able). Any preference before I get too far (i.e. separate function + tests or integrated in `brier_decomposition()` with tests)?

Reply:
I'm ok with either approach, but it sounds like it might be easier to test the separate function, why not go that way?

Reply:
Sounds good. I'll keep the separate function + tests approach.